addons/stock/models/stock_quant.py

```diff
@@ -599,8 +600,33 @@ def _gather(self, product_id, location_id, lot_id=None, package_id=None, owner_i
         removal_strategy = self._get_removal_strategy(product_id, location_id)
         removal_strategy_order = self._get_removal_strategy_order(removal_strategy)
         domain = self._get_gather_domain(product_id, location_id, lot_id, package_id, owner_id, strict)
-
-        return self.search(domain, order=removal_strategy_order).sorted(lambda q: not q.lot_id)
+        quants_cache = self.env.context.get('quants_cache')
+        if quants_cache is not None and strict:
+            res = self.env['stock.quant']
+            if lot_id:
+                res |= quants_cache[product_id.id, location_id.id, lot_id.id, package_id and package_id.id or False, owner_id and owner_id.id or False]
```

Comment:
I would use `.get` to avoid unwanted traceback

Reply:
It'a defaultdict